### PR TITLE
Reconciliation engine refactoring

### PR DESCRIPTION
**Pull request by tbak**

### CODE REVIEW

File: titus-common/src/main/java/com/netflix/titus/common/framework/reconciler/internal/SingleTransaction.java [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+    private final ReconciliationEngineMetrics<EVENT> metrics;
+    private final Clock clock;
+
+    private final ChangeAction changeAction;
+    private final Optional<Subscriber<Void>> changeActionSubscriber;
+    private final String transactionId;
+    private final long changeActionWaitTimeMs;
+    private final Subscription changeActionSubscription;
+
+    private TransactionStep transactionStep = TransactionStep.ChangeActionStarted;
```
Comment (fabiokung) [inline titus-common/src/main/java/com/netflix/titus/common/framework/reconciler/internal/SingleTransaction.java]: may need `volatile` here for visibility

File: titus-common/src/main/java/com/netflix/titus/common/framework/reconciler/internal/Transaction.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
+package com.netflix.titus.common.framework.reconciler.internal;
+
+import java.util.Optional;
+
+abstract class Transaction {
```
Comment (fabiokung) [inline titus-common/src/main/java/com/netflix/titus/common/framework/reconciler/internal/Transaction.java]: any reason this is an abstract class and not an interface?

Comment (fabiokung) [reply titus-common/src/main/java/com/netflix/titus/common/framework/reconciler/internal/Transaction.java]: is it to ensure methods themselves are not public?

File: titus-common/src/main/java/com/netflix/titus/common/framework/reconciler/internal/DefaultReconciliationEngine.java [hunk 3]
```diff
@@ -227,199 +201,114 @@ public EntityHolder getStoreView() {
...
-    }
-
-    private boolean hasRunningReconciliationActions() {
-        if (startedReconciliationActionSubscriptions.isEmpty()) {
-            return false;
-        }
-        boolean finished = startedReconciliationActionSubscriptions.stream().filter(a -> !a.isUnsubscribed()).count() == 0;
-        if (finished) {
-            startedReconciliationActionSubscriptions = Collections.emptyList();
+    private boolean isOverlapping(EntityHolder changePoint, List<EntityHolder> changePoints) {
```
[the remark this answers is not in the document]
Comment (tbak) [reply titus-common/src/main/java/com/netflix/titus/common/framework/reconciler/internal/DefaultReconciliationEngine.java]: The job will never be changed at this step. The reconciler drains the queue, only if the previous action(s) are completed, and models updated.

File: titus-common/src/main/java/com/netflix/titus/common/framework/reconciler/internal/DefaultReconciliationEngine.java [hunk 4]
```diff
@@ -227,199 +201,114 @@ public EntityHolder getStoreView() {
...
-                            }
-                    );
-            subscriptions.add(subscription);
+            emitEvent(eventFactory.newBeforeChangeEvent(this, changeAction, compositeTransactionId));
+
+            transactions.add(new SingleTransaction<>(this, changeAction, now, Optional.empty(), compositeTransactionId, true));
         }
-        this.startedReconciliationActionSubscriptions = subscriptions;
+
+        pendingTransaction = new CompositeTransaction(transactions);
```
[the remark this answers is not in the document]
Comment (tbak) [reply titus-common/src/main/java/com/netflix/titus/common/framework/reconciler/internal/DefaultReconciliationEngine.java]: Good point. I have fixed that.

File: titus-common/src/main/java/com/netflix/titus/common/framework/reconciler/internal/DefaultReconciliationFramework.java [hunk 5]
```diff
@@ -88,17 +89,22 @@
...
-                                          Function<EntityHolder, ReconciliationEngine<EVENT>> engineFactory,
+                                          Function<EntityHolder, InternalReconciliationEngine<EVENT>> engineFactory,
                                           long idleTimeoutMs,
                                           long activeTimeoutMs,
                                           Map<Object, Comparator<EntityHolder>> indexComparators,
                                           Registry registry,
                                           Optional<Scheduler> optionalScheduler) {
         Preconditions.checkArgument(idleTimeoutMs > 0, "idleTimeout <= 0 (%s)", idleTimeoutMs);
         Preconditions.checkArgument(activeTimeoutMs <= idleTimeoutMs, "activeTimeout(%s) > idleTimeout(%s)", activeTimeoutMs, idleTimeoutMs);
+        Preconditions.checkArgument(
```
[the remark this answers is not in the document]
Comment (tbak) [reply titus-common/src/main/java/com/netflix/titus/common/framework/reconciler/internal/DefaultReconciliationFramework.java]: I changed `JobReconciliationFrameworkFactory` to pass `InternalReconciliationEngine`. We need to cleanup this later, and properly hide `InternalReconciliationEngine`.

File: titus-common/src/main/java/com/netflix/titus/common/framework/reconciler/internal/DefaultReconciliationFramework.java [hunk 6]
```diff
@@ -125,10 +131,12 @@ public DefaultReconciliationFramework(List<ReconciliationEngine<EVENT>> bootstra
         this.internalEventSubscription = eventsObservable.subscribe(ObservableExt.silentSubscriber());
 
         this.loopExecutionTime = registry.timer(LOOP_EXECUTION_TIME_METRIC);
+        this.lastFullCycleExecutionTimeMs = scheduler.now() - idleTimeoutMs;
         this.lastExecutionTimeMs = scheduler.now();
         PolledMeter.using(registry).withName(LAST_EXECUTION_TIME_METRIC).monitorValue(this, self -> scheduler.now() - self.lastExecutionTimeMs);
+        PolledMeter.using(registry).withName(LAST_FULL_CYCLE_EXECUTION_TIME_METRIC).monitorValue(this, self -> scheduler.now() - self.lastFullCycleExecutionTimeMs);
 
-        engines.addAll(bootstrapEngines);
+        engines.addAll(bootstrapEngines.stream().map(e -> (InternalReconciliationEngine<EVENT>) e).collect(Collectors.toList()));
```
[the remark this answers is not in the document]
Comment (tbak) [reply titus-common/src/main/java/com/netflix/titus/common/framework/reconciler/internal/DefaultReconciliationFramework.java]: I have simplified this as recommended.

File: titus-common/src/main/java/com/netflix/titus/common/framework/reconciler/internal/Transaction.java (hunk 2 above)
Comment (tbak) [reply titus-common/src/main/java/com/netflix/titus/common/framework/reconciler/internal/Transaction.java]: Initially I planned to have common skeleton implementation, but it ended up as a plain interface. I will change that.

File: titus-common/src/main/java/com/netflix/titus/common/framework/reconciler/internal/DefaultReconciliationFramework.java [hunk 7]
```diff
@@ -254,70 +266,89 @@ private void doSchedule(long delayMs) {
         }, delayMs, TimeUnit.MILLISECONDS);
     }
 
-    private long doLoop() {
+    private void doLoop(boolean fullReconciliationCycle) {
+        Set<InternalReconciliationEngine<EVENT>> mustRunEngines = Collections.newSetFromMap(new HashMap<>());
```
Comment (corindwyer) [inline titus-common/src/main/java/com/netflix/titus/common/framework/reconciler/internal/DefaultReconciliationFramework.java]: use `new HashSet`

Comment (tbak) [reply titus-common/src/main/java/com/netflix/titus/common/framework/reconciler/internal/DefaultReconciliationFramework.java]: Fixed

**Pull request merged**